festim/generic_simulation.py

```diff
@@ -208,13 +208,32 @@ def attribute_boundary_conditions(self):
             + [i + 1 for i, _ in enumerate(self.traps.traps)]
         )
 
+        # collect all DirichletBCs
+        dc_bcs = [
+            bc
+            for bc in self.boundary_conditions
+            if isinstance(bc, festim.DirichletBC) and (bc.field in valid_fields)
+        ]
+
         for bc in self.boundary_conditions:
             if bc.field not in valid_fields:
                 raise ValueError(f"{bc.field} is not a valid field for BC")
             if bc.field == "T":
                 self.T.boundary_conditions.append(bc)
             else:
                 self.h_transport_problem.boundary_conditions.append(bc)
+            # checks that DirichletBC is not set with another bc on the same surface
+            # iterate through all BCs
+            for dc_bc in dc_bcs:
+                if (
+                    bc == dc_bc or bc.field != dc_bc.field
+                ):  # skip if the same BC or different fields
+                    continue
+                for surf in bc.surfaces:
+                    if surf in dc_bc.surfaces:
+                        raise ValueError(
+                            f"A DirichletBC is simultaneously set with another boundary condition on surface {surf} for field {dc_bc.field}"
+                        )
```

Comment:
I also wonder if we can speed up this a bit, since its a nested loop. I don't know how complicated simulations might be, but another possibility is to use the `set` method as mentioned [here](https://stackoverflow.com/questions/3170055/test-if-lists-share-any-items-in-python). The modified part would be:
```python
for dc_bc in dc_bcs:
    if (
        bc == dc_bc or bc.field != dc_bc.field
    ):  # skip if the same BC or different fields
        continue
    intersection = set(bc.surfaces) & set(dc_bc.surfaces)
    if bool(intersection):
        raise ValueError(
            f"A DirichletBC is simultaneously set with another boundary condition on surfaces {intersection} for field {bc.field}"
        )
                    )
``` 
@RemDelaporteMathurin, please, let me know what do you think

Reply:
Sure why not! Just make sure to add enough comments to describe what is done here cause this may look complex for non-python experts